### DISCUSSION TITLE
delete键原生操作怎么实现

### DISCUSSION THREAD

**Discussion by chaoyue1217** [opensumi/core#2287] [Q&A] [answered]
重新注册delete键的功能：当按下 delete 键时，根据当前所处的环境，如果是input或者textArea，则执行delete原生操作，如果是处于文档中，则执行自定义的操作。当使用`commandService.executeCommand()`还原原生操作时，electron.redo/undo/copy 等都存在，但是没有elector.delete，这个delete功能要怎么去实现？

**Comment by erha19** [comment, marked as answer]
@chaoyue1217 目前 Electron 相关菜单可以通过 `nativeRole`方式去指定原生操作，相关代码见：https://github.com/opensumi/core/blob/068786d0cbdd753b8b7597a9dca386400f89b310/packages/core-browser/src/common/common.contribution.ts#L239~L278

看一下是否能解决你的问题？

**Comment by erha19** [comment]
@chaoyue1217  你如果希望实现这种快捷键逻辑，你可以通过追加 `when` 条件的方式来进行快捷键的注册，相对会复杂一点，目前，你可以先试一下在文档内按快捷键的命令相应，相关代码可以断点查看一下这里：

https://github.com/opensumi/core/blob/670f4f48a72b2cd827cba6c7d9d87c3c158eed4b/packages/core-browser/src/keybinding/keybinding.ts#L882

这里输出的信息代表了你按下一个快捷键响应的命令，以及对应的执行优先级，目前执行快捷键命令的优先级为:

第一优先级：带 when 表达式的优先级大于不带 when 表达式的
第二优先级：priority 参数

处于文本文档中的 when 表达式可用 `editorFocus` , 编辑器组件中可用 `inEditorComponent` ，原生情况下默认处理，或者自己注册 `when` 表达式进行维护也可以

参考文档：https://opensumi.com/zh/docs/integrate/universal-integrate-case/custom-keybinding#%E9%80%9A%E8%BF%87-when-%E6%8E%A7%E5%88%B6%E7%94%9F%E6%95%88%E8%8C%83%E5%9B%B4

**Reply by chaoyue1217** [reply]
我希望通过调用`commandService.executeCommand`方法直接去调用原生操作。

我在input中测试到` ctrl + delete `是可以用的，在【快捷键设置】里找到了 对应的命令为 `deleteWordRight` 如下图：
![image](https://user-images.githubusercontent.com/18662948/218916145-89001843-e36b-4fd4-bf3e-748dcd9755c9.png)

但是当执行`commandService.executeCommand('deleteWordRight')`时，没有作用，命令换成别的，如`editor.close` 是执行了对应操作的。 为什么`deleteWordRight `没有被执行呢？

**Reply by erha19** [reply]
你可以按照我上面说到的断点位置，断点查看一下快捷键按下时的具体执行路径排查问题。https://github.com/opensumi/core/blob/670f4f48a72b2cd827cba6c7d9d87c3c158eed4b/packages/core-browser/src/keybinding/keybinding.ts#L882
